diff --git a/data/org.gnome.Evolution.appdata.xml.in.in b/data/org.gnome.Evolution.appdata.xml.in.in
index f6c7911aa2b12d9a74d71e13c2c142ac9be295f5..2809482f98dc04bd69dd14fcf230284eaba36082 100644
--- a/data/org.gnome.Evolution.appdata.xml.in.in
+++ b/data/org.gnome.Evolution.appdata.xml.in.in
@@ -26,9 +26,6 @@
   <url type="vcs-browser">https://gitlab.gnome.org/GNOME/evolution</url>
   <url type="translate">https://l10n.gnome.org/module/evolution/</url>
   <launchable type="desktop-id">org.gnome.Evolution.desktop</launchable>
-  <provides>
-    <id>org.gnome.Evolution.desktop</id>
-  </provides>
   <branding>
     <color type="primary" scheme_preference="light">#d6dfd9</color>
     <color type="primary" scheme_preference="dark">#613583</color>
